### ISSUE TITLE
内存寻址 | 始终

### ISSUE THREAD

**Issue by Liam0205** [Liam0205/liam0205.github.io#111]
https://liam.page/2016/05/01/Introduction-to-Memory-Addressing/ 

 公认的计算机之父是阿兰·图灵在 1936 年提出了「图灵机」的概念。图灵机是一个抽象的计算模型，它可以形象地表述为：  有一个向两端无限延伸的带子，可以在上面记录内容 有一个可以在带子上擦除/写入的读写头，可以在带子上任意移动 有一个控制器，控制读写头的移动和擦除/写入操作  图灵机的表述看似纸上谈兵，但实际与现在计算机的体系结构有良好的对应，甚至可以说是所有现代计算机背后的灵魂。  带子对应于内

**Comment by randomwangran** [comment]
good point! i've learning the memory and struggle it right now. Hope your blog can save me.

**Comment by Liam0205** [comment]
@randomwangran I believe that it'll do help.

**Comment by pAXkkkxh** [comment]
>显而易见，最简单的方式是建立一张足够大的表（称为页表）：表中的 key 由线性地址的首 20 位计算得到（或者干脆就是它自己）；表中的 value 则是对应页帧在物理内存中的起始地址（帧基址）。

>这种情况下，我们假设极端的情况：每个进程都请求了 4 GiB 的线性空间。按照 4 KiB 每页计算，共需 1 M 个页帧信息。每份页帧信息至少包括 20 位的帧基址，再加上其他信息凑齐 4 Bytes。

有个疑问，每个页表项的key是20位或者32位，value是帧基址20位，再加上其他信息已经超过4 Bytes了啊？

**Comment by Liam0205** [comment]
> @labixiaoshuai 
> 有个疑问，每个页表项的key是20位或者32位，value是帧基址20位，再加上其他信息已经超过4 Bytes了啊？

key 本身不会保存在页表项里。

**Comment by Liam0205** [comment]
> @randomwangran
>      I am not sure the correct jargon here. Linear memory is just
>      another way to manage memory? For the "physical memory", the
>      address of it is this: 1,2,3,4,5,6,...
>      
>      For the "linear memory", we say memory address is like this:
>      1,3,5,7,9...
> 
>      The first element of both "physical" and "linear memory" is 1,
>      but the second is different. "Physical" memory is 2, whereas
>      "linear" memory is 3.
> 
>      I am correct?

No, you're not.

For memory address, both that of linear memory and physical memory are 
continuous. And there is a mapping between linear memory and physical 
memory.

The mapping allows multiple processes share physical memory, and, at the
same time, each process has a linear continuous address.

**Comment by pAXkkkxh** [comment]
386的段描述符中，为什么不把段基址和段界限连续放呢？段表和页表是放在虚拟地址空间的内核空间中吗？这部分内存反映到实际物理内存是在高1G的部分吗？

**Comment by Liam0205** [comment]
> @labixiaoshuai
> 386的段描述符中，为什么不把段基址和段界限连续放呢？段表和页表是放在虚拟地址空间的内核空间中吗？这部分内存反映到实际物理内存是在高1G的部分吗？

好问题。

具体原因我也不太清楚，这个没有深入去探究。
但是我的看法是，大概率是要结合发展的眼光来考虑。很有可能一开始只用到了 32bit，所以你看最低 32bit 里保存的是连续的段基址和段界限。后来增加了各种字段，也延长了段基址和段界限的 bit 数，所以就不连续了。

**Comment by tinyMai** [comment]
您好，我提出3个疑惑：
* "每个进程在一级页表中对应一个表项"是否打错了，应该是"每个进程对应一个一级页表"，一个进程对应一个一级页表，一个页表对应1k_1个二级页表，一个二级页表对应1k_2个页框(4kb)，所以产生 1k_1 * 1k_2 * 4kb = 4GB 的分配空间。
* 对于386的寻址能力的计算应该是错的，比如说实模式下，段有 2^16 个，段大小是 2^16 B, 但是地址空间不是 2^16 * 2^16B，地址空间的决定性因素是地址的长度，实模式20位地址的空间大小是 2^20 B。
* 硬编码的概念是否能用到实模式寻址上...

期待您的回复，感谢！

**Comment by a1mersnow** [comment]
32 位的线性地址只能寻址 2 的 32 次方的空间吧，也就是4G，怎么就变成 64T 了呢？这个地方真的很难理解。前边再怎么分段，和偏移地址组合起来还是 32 位的线性地址啊…… 大佬麻烦解答下
